runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/BoundedTrieData.java

```diff
@@ -0,0 +1,585 @@
+
+  private static final int DEFAULT_BOUND = 100; // Default maximum size of the trie
+
+  /** Returns the singleton path if this {@link BoundedTrieData} represents a single path. */
+  @Nullable private List<String> singleton;
+
+  /**
+   * Returns an {@link Optional} containing the root {@link BoundedTrieNode} if this {@link
+   * BoundedTrieData} represents a trie.
+   */
+  @Nullable private BoundedTrieNode root;
+
+  /** Returns the maximum size of the trie. */
+  private int bound;
+
+  public BoundedTrieData() {
+    this(null, null, DEFAULT_BOUND);
+  }
+
+  public BoundedTrieData(List<String> singleton) {
+    this(singleton, null, DEFAULT_BOUND);
+  }
+
+  public BoundedTrieData(BoundedTrieNode root) {
+    this(null, root, DEFAULT_BOUND);
+  }
+
+  /**
+   * Constructs a new BoundedTrieData object.
+   *
+   * <p>A BoundedTrieData object represents the data stored in a bounded trie. It can either be a
+   * singleton list of strings, or a trie with a given root and bound.
+   *
+   * @param singleton the singleton list of strings, or null if the data is a trie
+   * @param root the root of the trie, or null if the data is a singleton list
+   * @param bound the maximum number of elements allowed in the trie
+   * @throws IllegalArgumentException if both {@code singleton} and {@code root} are non-null
+   */
+  public BoundedTrieData(
+      @Nullable List<String> singleton, @Nullable BoundedTrieNode root, int bound) {
+    assert singleton == null || root == null;
+    this.singleton = singleton;
+    this.root = root;
+    this.bound = bound;
+  }
+
+  /** Converts this {@link BoundedTrieData} to its proto {@link BoundedTrie}. */
+  public synchronized BoundedTrie toProto() {
+    BoundedTrie.Builder builder = BoundedTrie.newBuilder();
+    builder.setBound(this.bound);
+    if (this.singleton != null) {
+      builder.addAllSingleton(this.singleton);
+    }
+    if (this.root != null) {
+      builder.setRoot(this.root.toProto());
+    }
+    return builder.build();
+  }
+
+  /** Creates a {@link BoundedTrieData} instance from its proto {@link BoundedTrie}. */
+  public static BoundedTrieData fromProto(BoundedTrie proto) {
+    if (proto.hasRoot()) {
+      return new BoundedTrieData(
+          null, BoundedTrieNode.fromProto(proto.getRoot()), proto.getBound());
+    } else {
+      return new BoundedTrieData(proto.getSingletonList(), null, proto.getBound());
+    }
+  }
+
+  /** Returns this {@link BoundedTrieData} as a {@link BoundedTrieNode}. */
+  @Nonnull
+  private synchronized BoundedTrieNode asTrie() {
+    if (this.root != null) {
+      return this.root;
+    } else {
+      BoundedTrieNode trieNode = new BoundedTrieNode();
+      if (this.singleton != null) {
+        trieNode.add(this.singleton);
+      }
+      return trieNode;
+    }
+  }
+
+  /** Returns a new {@link BoundedTrieData} instance that is a deep copy of this instance. */
+  public synchronized BoundedTrieData getCumulative() {
```

Comment:
This returns a BoundedTrieData over a Set<List> to be consistent with other MetricData in java SDK